TELCODOCS-1576: Image-based upgrade with Lifecycle Agent (#69809)
openshift/openshift-docs: modules/ztp-image-based-upgrade.adoc

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} cluster with the {lcao}

> The {lcao} provides you with an alternative way to upgrade the platform version of a managed {sno} cluster.

Reviewer: remove managed

> The image-based upgrade is faster than the standard upgrade method and allows you to upgrade from {product-title} <4.y> to <4.y+2> and <4.y.z> to <4.y.z+n>.

Reviewer 2: nit - maybe consistency with the format of <4.y**+2**> <4.y.z**+n**>. Perhaps +n is more generic

Author: What I'm trying to say here is that y-stream releases are limited to 2 releases (upgrade from 4.14 to 4.16 is ok, but not to 4.18) and there's no limit to how many z-stream upgrades you can do (upgrade from 4.14.10 to 4.14.30 is okay). Hence the `n`. If it needs this much explanation, it's not clear enough

Reviewer 2: Oh, I thought +2 was just an example. Perhaps it's just me?

Author: No, it could be that I need to rephrase. WDYT about this?
> ...allows you to directly upgrade from {product-title} <4.y> to <4.y+2>, and <4.y.z> to <4.y.z+n>

Does that `directly upgrade` help to emphasize what we're trying to introduce?

Reviewer 2: Yeah that helps i think

This upgrade method utilizes generated images from a dedicated seed cluster.

> A seed cluster is a managed {SNO} deployed deployed and configured with the target {product-title} version and applications.

Reviewer 2: Typo "deployed deployed"

> A seed cluster is a managed {sno} cluster deployed and configured with the target {product-title} version and applications.

Reviewer: remove managed

> A seed cluster is a managed {sno} cluster deployed and configured with the target {product-title} version and applications.

Reviewer: Applications would generally not be part of the seed image. The seed image would include: 
- The target OCP version 
- The day 2 operators
- Configuration that is common to all target SNOs that the seed will be used for

Author: @browsell I revised this based on the linked section. Please let me know if the current level of detail is okay or if you want to see more.

> You can use the seed image, which is generated from the seed cluster, to upgrade the platform version on any {sno} that has the same combination of hardware, Day 2 Operators, and cluster configuration as the seed cluster.

Reviewer 2: Should it be "{sno} cluster"

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform that the clusters are running on.
Each different hardware platform requires a separate seed image.
====
// TODO agree on how much we should specify this. Is this enough or detail about CPU topology, deployment method, etc.

The {lcao} uses two custom resources (CRs) on the participating clusters:

* On the seed cluster, the `SeedGenerator` CR allows for the seed image generation. This CR specifies the repository to push the seed image to.

> * On the managed target cluster, the `ImageBasedUpgrade` CR specifies the seed container image for the upgrade of the target cluster and the backup configurations for your workloads.

Reviewer: Remove managed

> // Module included in the following assemblies:

Reviewer: We should add some more detail around backup and restore along with extra manifest, usecases etc. This doc may be a good starting point https://github.com/openshift-kni/lifecycle-agent/blob/main/docs/image-based-upgrade.md

// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} cluster with the {lcao}

You can upgrade the platform version of a managed {sno} cluster with the {lcao}.

> Using a generated seed image allows you to skip multiple minor releases and it is faster than the standard upgrade method.

Reviewer: Get PM input on the upgrade paths that will be supported

> Using a generated seed image allows you to skip up to two minor releases and it is faster than the standard upgrade method.

Reviewer 3: I would start with some context, saying this operator brings an alternative way of managing upgrades on SNO. Alternative to the standard upgrades method (with a link). And I suggest point first to the main features, it does it much more faster to satisfy usual upgrade requirements on Telco. And also, that you can do OCP upgrades x.y+2 seamless. In the first paragraph maybe is not needed to talk about seed images.
Keeping the things about seed images, that are implementation details, for the second paragraph.

> You generate the seed image from a {sno} seed cluster, which is deployed and configured with the desired {product-title} and applications.

Reviewer 3: maybe we would give a little bit more detail on what is a "seed cluster". We have a hub cluster, then we have any spokes cluster, and now a new concept "seed cluster". Which is out of the ZTP hub/spokes clusters. 
Then we can refer to this definition on other parts of the documentation.

You can then use that seed image to upgrade the platform version on any {sno} that has the same combination of hardware, Day 2 Operators, and cluster configuration as the seed cluster.

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform that the clusters are running on.
Each different hardware platform requires a separate seed image.
====
// TODO agree on how much we should specify this. Is this enough or detail about CPU topology, deployment method, etc.

The {lcao} uses two custom resources (CRs) on the participating clusters:

> * On the seed cluster, the `SeedGenerator` CR allows the seed image generation. This CR specifies the repository to push the seed image to.

Reviewer 4: missing a word? `allows for(?) the seed image generation`?

> * On the seed cluster, the `SeedGenerator` CR allows the seed image generation. This CR specifies the repository to push the seed image to.

Reviewer 3: Here, the "seed cluster" have a better picture if we could have a definition above. Maybe with a pic.

> * On the target cluster, the `ImageBasedUpgrade` CR specifies the seed container image for the upgrade of the target cluster and the backup configurations for your workloads.

Reviewer 3: would make sense to say: "target spoke cluster" ?

Reviewer: nit, use managed not spoke

.Example SeedGenerator CR
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage
spec:
  seedImage: <seed_container_image>
----

.Example ImageBasedUpgrade CR
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: ImageBasedUpgrade
metadata:
  generation: 1
  name: upgrade
spec:
  stage: Idle

> status:

Reviewer 4: I don't think we normally include the `status` field when providing example CRs

  conditions:
  - lastTransitionTime: "2024-01-19T05:42:58Z"
    message: Idle
    observedGeneration: 1
    reason: Idle
    status: "True"
    type: Idle
  observedGeneration: 1
----

After generating the seed image on the seed cluster, you can move through the stages on the target cluster by specifying the image to pull and then setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:

* `Idle`
* `Prep`
* `Upgrade`
* `Rollback` (Optional)

> * `Idle`

Reviewer 4: `Idle` is listed twice?

Author: There was a few iterations to this because you start in Idle and end the process in Idle. I'll delete the extra one and see if I can add a diagram to demonstrate the flow better.

* `Idle`

[discrete]
== Idle stage

This is the default stage when the {lcao} is installed. There is no ongoing upgrade and the cluster is ready to move to the `Prep` stage.

> After a successful upgrade or a rollback, you commit to the change by manually moving the stage to `Idle`.

Reviewer: nit, instead of manual, patch spec stage to "Idle"

[discrete]
== Prep stage

> You can complete this stage before a scheduled maintenance window.

Reviewer 4: maybe make this a note or call out rather than inline?

You can complete this stage before a scheduled maintenance window.

In the `Prep` stage, the {lcao} ensures that the target cluster is ready to proceed to the `Upgrade` stage by checking if it meets certain conditions.
The Operator uses the `ostree` versioning system and the OADP Operator to create a backup, which allow complete cluster reconfiguration after upgrade or rollback.

At the `Prep` stage, the {lcao} pulls the seed image and additional container images specified in the seed image to the target cluster.

You can stop the upgrade process.

> If you stop, the Operator performs cleanup operations and sets the upgrade stage back to `Idle`.

Reviewer: To stop/abort the upgrade the stage must be patched back to "Idle"

[discrete]
== Upgrade stage

> Before the {lcao} starts the upgrade process, you can create an optional backup of your cluster resources with the OADP Operator.

Reviewer 3: In the Prep stage (:81) also says about preparing backup. Where is the backup really done?

Reviewer 3: ok, reading further, it seems that on "prep" you specify what do you want to backup. Then on "upgrade" the backup is done. Would this be more clear in the definition of both stages?

Before the {lcao} starts the upgrade process, you can create an optional backup of your cluster resources with the OADP Operator.
The OADP Operator creates a backup of the user-specified resources on a compatible Object storage solution, which are defined in the `Backup` CR.

> After the target cluster reboots with the new platform version, the Operator applies the cluster and application configurations defined in the `Restore` CR.

Reviewer: It also applies any extra manifests specified in the configmap referenced in the spec

> The Operator also regenerates the cluster certificate.

Reviewer: regenerates cluster certificates.

This ensures that each {sno} cluster upgraded with the same seed image has unique and valid cryptographic objects.

[discrete]
== (Optional) Rollback stage

> The rollback stage can be initiated manually or automatically.

Reviewer: Should elaborate on the automatic rollback. This is initiated if the LCA has detected the upgrade has failed or is taking longer than the specified limit.

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} cluster with the {lcao}

You can upgrade the platform version of a managed {sno} cluster with the {lcao}.
Using a generated seed image allows you to skip multiple minor releases and it is faster than the standard upgrade method.

> You generate the seed image from a {sno} seed cluster, which is deployed and configured with the desired {product-title} and applications. You can upgrade the platform version on any {sno} that has the same combination of hardware, Day 2 Operators, and cluster configuration.

Reviewer 5: ```suggestion
You generate the seed image from a {sno} seed cluster, which is deployed and configured with the desired {product-title} and applications. You can upgrade the platform version on any {sno} that has the same combination of hardware, Day 2 Operators, and cluster configuration as the seed cluster.
```

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform that the clusters are running on.

> Each different hardware platform requires a separate custom image.

Reviewer 5: ```suggestion
Each different hardware platform requires a separate seed image.
```

====
// TODO agree on how much we should specify this. Is this enough or detail about CPU topology, deployment method, etc.

The {lcao} uses two custom resources (CRs) on the participating clusters:

* On the seed cluster, the `SeedGenerator` CR allows the seed image generation. This CR specifies the repository to push the seed image to.
* On the target cluster, the `ImageBasedUpgrade` CR specifies the seed container image for the upgrade of the target cluster and the backup configurations for your workloads.

.Example SeedGenerator CR
+
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage
spec:
  seedImage: <seed_container_image>
----

.Example ImageBasedUpgrade CR

> +

Reviewer 5: This `+` is rendered weirdly

![image](https://github.com/openshift/openshift-docs/assets/10882062/3c4c9d4f-b825-4a3f-8735-8d70bcd906fd)

[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: ImageBasedUpgrade
metadata:
  generation: 1
  name: upgrade
spec:
  stage: Idle
status:
  conditions:
  - lastTransitionTime: "2024-01-19T05:42:58Z"
    message: Idle
    observedGeneration: 1
    reason: Idle
    status: "True"
    type: Idle
  observedGeneration: 1
----

> After generating the seed image on the seed cluster, you can move through the stages on the target cluster by setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:

Reviewer 5: ```suggestion
After generating the seed image on the seed cluster, you can move through the `upgrade` stages on the target cluster by <specifying the pull-spec of the seed image and then> setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:
```

* `Idle`
* `Prep`
* `Upgrade`
* `Rollback` (Optional)

[discrete]
== Idle stage

This is the default stage when the {lcao} is installed. There is no ongoing upgrade and the cluster is ready to move to the `Prep` stage.

After a successful upgrade or a rollback, you commit to the change by patching the `stage` field to `Idle` in the `ImageBasedUpgrade` CR.
Changing to this stage ensures that the {lcao} cleans up resources, so the cluster is ready for upgrades again.

[discrete]
== Prep stage

[NOTE]
====
You can complete this stage before a scheduled maintenance window.
====

During the `Prep` stage, you specify the details of the upgrade and the {lcao} prepares for the upgrade without impacting the current running version.
This preparation includes ensuring that the target cluster is ready to proceed to the `Upgrade` stage by checking if it meets certain conditions and pulling the seed image to the target cluster with additional container images specified in the seed image.

> You can prepare backup resources with the OADP Operator's `Backup` and `Restore` CRs.

Author: I'm cleaning this up so backing up is not marked as optional. Theoretically it is optional, but after checking with devs, it looks like that it's not probable that customers would want to skip creating backup.

> These CRs are used in the `Upgrade` stage to reconfigure the cluster, register the cluster with {rh-rhacm} if applicable, and restore application artifacts.

Author: This refers to the fact that you can do the IBU process without ACM. However, I realize that in the current state - and the missing flow for that particular use case - it is causing more confusion, so I'm going to clean this up for now

#  oadpContent: <9>
#  - name: oadp-cm-example
#    namespace: openshift-adp
----
<1> Defines the desired stage for the `ImageBasedUpgrade` CR. The value can be `Idle`, `Prep`, `Upgrade`, or `Rollback`.
<2> Defines the target platform version, the seed image to be used, and the secret required to access the image.
<3> Configures the automatic rollback. By default, automatic rollback on failure is enabled throughout the upgrade.
<4> (Optional) Disables automatic rollback when the reconfiguration of the cluster fails upon the first reboot.
<5> (Optional) Disables automatic rollback after the Lifecycle Agent Operator reports a failed upgrade upon completion.
<6> (Optional) Disables automatic rollback when the upgrade does not complete after reboot within the time frame specified in the `initMonitorTimeoutSeconds` field.
<7> (Optional) Specifies the time frame in seconds. If not defined, the default value is 1800 seconds or 30 minutes.
<8> (Optional) Specify the list of `ConfigMap` CRs that contain the additional extra manifests that you want to apply to the target cluster.
<9>	(Optional) Specify the list of `ConfigMap` CRs that contain the OADP `Backup` and `Restore` CRs.

> After generating the seed image on the seed cluster, you can move through the stages on the target cluster by specifying the image to pull and then setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:

Reviewer 6: It's not possible to move between some stages.
For example:
   can't move from Upgrade to Prep.
   can't move from Rollback to Prep.

Do we want to mention it.

* `Idle`
* `Prep`
* `Upgrade`
* `Rollback` (Optional)

[discrete]
== Idle stage

This is the default stage when the {lcao} is installed. There is no ongoing upgrade and the cluster is ready to move to the `Prep` stage.

After a successful upgrade or a rollback, you commit to the change by patching the `stage` field to `Idle` in the `ImageBasedUpgrade` CR.

> Changing to this stage ensures that the {lcao} cleans up resources, so the cluster is ready for upgrades again.

Reviewer 6: Maybe we should a warning here, that moving to idle after rollback - removes the respective ostree (would be otherwise used for debugging the failed upgrade)

[IMPORTANT]
====
The same version of the applications must function on both the current and the target release of {product-title}.
====

Additionally to the OADP Operator, the {lcao} uses the `ostree` versioning system to create a backup, which allow complete cluster reconfiguration after both upgrade and rollback.

You can stop the upgrade process at this point by patching the `stage` field to `Idle` in the `ImageBasedUpgrade` CR.
If you stop, the Operator performs cleanup operations.

[discrete]
== Upgrade stage

Just before the {lcao} starts the upgrade process, the backup of your cluster resources specified in the `Prep` stage are created on a compatible Object storage solution.
After the target cluster reboots with the new platform version, the Operator applies the cluster and application configurations defined in the `Backup` and `Restore` CRs, and applies any extra manifests that are specified in the referenced `ConfigMap` CR.

The Operator also regenerates cluster certificates.
This ensures that each {sno} cluster upgraded with the same seed image has unique and valid cryptographic objects.

[discrete]
== (Optional) Rollback stage

> The rollback stage can be initiated manually or automatically.

Reviewer 6: automatically by default upon failure

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} cluster with the {lcao}

From {product-title} 4.14.7, the {lcao} 4.15 provides you with an alternative way to upgrade the platform version of a {sno} cluster.

> The image-based upgrade is faster than the standard upgrade method and allows you to directly upgrade from {product-title} <4.y> to <4.y+2>, and <4.y.z> to <4.y.z+n>.

Reviewer 7: ```suggestion
The image-based upgrade is faster than the standard upgrade method and allows you to directly upgrade from an {product-title} minor version to a minor version two releases newer, for example from 4.13 to 4.15.
You can also use image-based upgrades to upgrade between z-stream releases, for example, from 4.15.1 to 4.15.10. 
```

Reviewer 8: if specifying examples, I'd suggest using versions where LCA is supported / available in official catalogs to avoid any possible confusion

`for example from 4.13 to 4.15` maybe better to `for example from 4.14 to 4.16`

Reviewer 7: Yes wasn't sure about supported versions thanks

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} clusters with the {lcao}

You can upgrade the platform version on a fleet of managed {sno} clusters with the {lcao}.
This upgrade process allows you to skip multiple minor releases and it is faster than the standard upgrade method by using generated seed images.
With the seed image, you can upgrade the platform version on any {sno} that matches the image's cluster profile.

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform that a fleet of clusters is running on.
Each different hardware platform requires a separate custom image.
====

After generating the seed image, the upgrade process has four stages:

`Idle`:: There is no ongoing upgrade, the cluster is ready to move to the preparation stage.

> `Prep`:: The {lcao} ensures that the the node is ready to proceed to the next stage and {cgu-operator} applies the necessary {rh-rhacm} policies before the upgrade. You can do this well in advance and you can stop the upgrade process at this point.

Reviewer 7: ```suggestion
`Prep`:: The {lcao} ensures that the node is ready to proceed to the next stage and {cgu-operator} applies the necessary {rh-rhacm} policies before the upgrade. You can do this well in advance and you can stop the upgrade process at this point.
```

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} cluster with the {lcao}

You can upgrade the platform version of a managed {sno} cluster with the {lcao}.
Using a generated seed image allows you to skip multiple minor releases and it is faster than the standard upgrade method.

> You generate the seed image from a seed cluster, which is deployed and configured with the desired {product-title} and applications. You can upgrade the platform version on any {sno} that matches the cluster profile of the seed image.

Reviewer 5: ```suggestion
You generate the seed image from a seed single-node cluster, which is deployed and configured with the desired {product-title} and applications. You can upgrade the platform version on any {sno} that matches the cluster profile of the seed image.
```

The word cluster inherently implies multiple nodes, so mentioning it without "single-node" could give the impression that "Seed cluster" refers to the hub, or something else, that is not single-node.

> You generate the seed image from a seed cluster, which is deployed and configured with the desired {product-title} and applications. You can upgrade the platform version on any {sno} that matches the cluster profile of the seed image.

Reviewer 5: >cluster profile

Cluster profile is an internal technical term we use to refer to the combination of hardware / day-2 operators / cluster configuration that is desired for a group of clusters that are to-be-upgraded, it is not an official term recognized by Red Hat / OpenShift and so it should preferably be avoided or defined

Author: Thanks! This is the only part I used cluster profile, so I just provide an explanation as to what needs to match (combination of hardware / day-2 operators / cluster configuration).

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform that the clusters are running on.
Each different hardware platform requires a separate custom image.
====
// TODO agree on how much we should specify this. Is this enough or detail about CPU topology, deployment method, etc.

The {lcao} uses two custom resources (CRs):

> * The `SeedGenerator` CR specifies the repository to pull the seed image from.

Reviewer 5: ```suggestion
* The `SeedGenerator` CR specifies the repository to push the seed image to.
```

> * The `ImageBasedUpgrade` CR specifies the seed container image for the upgrade of the target cluster and the OADP `Backup` and `Restore` CR configurations to back up your workloads.

Reviewer 5: {lcao} runs in two different contexts:

* On the seed cluster, where it will reconcile the SeedGenerator and generate a seed image

* On the cluster-to-be-upgraded, where it will reconcile the ImageBasedUpgrade by pulling a seed image and upgrading from it.

I don't feel these bullets and the line above them communicate that well

Author: I rephrased this, PTAL.

> After generating the seed image, you can move throught the stages by setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:

Reviewer 5: ```suggestion
After generating the seed image {on the seed cluster...}, you can {on the cluster to be upgraded...} move throught the stages by setting the `spec.stage` field to the following values in the `ImageBasedUpgrade` CR:
```

After a successful upgrade or a rollback, you commit to the change by manually moving the stage to `Idle`.
Changing to this stage ensures that the {lcao} cleans up resources, so the cluster is ready for upgrades.

[discrete]
== Prep stage

You can complete this stage before a scheduled maintenance window.

In the `Prep` stage, the {lcao} ensures that the target clusters are ready to proceed to the `Upgrade` stage by checking if they meet certain conditions.
The Operator uses the `ostree` versioning system and the OADP Operator to create a backup, which allow complete cluster reconfiguration after upgrade or rollback.

At the `Prep` stage, the {lcao} pulls the seed image and additional container images specified in the seed image.

You can stop the upgrade process. If you stop, the Operator performs cleanup operations and sets the upgrade stage back to `Idle`.

[discrete]
== Upgrade stage

Just before the {lcao} starts the upgrade process, the OADP Operator creates a backup of the user-specified resources on a compatible Object storage solution, such as an Amazon S3 bucket.

> These OADP `Backup` and `Restore` CRs are applied to the target cluster after the upgrade.

Reviewer 9: Well, the backup is done during the upgrade stage before booting in the new stateroot. But the recover CR is executed once the new version of the SNO is ready, and once the platform is ready, then recover the apps.

Author: I rearranged and rephrased this part, also emphasizing that backup is optional. WDYT?
> Before the {lcao} starts the upgrade process, you can create an optional backup of your cluster resources with the OADP Operator.
The OADP Operator creates a backup of the user-specified resources on a compatible Object storage solution, which are defined in the `Backup` CR.
After the target cluster reboots with the new platform version, the Operator applies the cluster and application configurations defined in the `Restore` CR.

Once the upgrade is successful, the target cluster boots with the new platform version, then the Operator applies the cluster and application configurations.

The Operator also regenerates the cluster certificate.
This ensures that each {sno} cluster upgraded with the same seed image has unique and valid cryptographic objects.

[discrete]
== Rollback stage (Optional)

> You can use the backup from the `Upgrade` stage to manually start a rollback if the upgrade fails.

Reviewer 9: Maybe it is confusing by using the word backup. Do you mean the backup created by OADP or the previous stateroot? For rollback it just boots in the previous stateroot, basically setting it as the default boot.

Author: I meant the OADP backup but now I realize that I still confuse when backup vs previous stateroot is used.
To avoid mentioning stateroots, I changed it like this:
> You can manually start a rollback if the upgrade fails.
> If you roll back the upgrade, the cluster boots up with the previous release of {product-title} and application configurations.

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]
= Image-based upgrade for {sno} clusters with the {lcao}

You can upgrade the platform version of the managed {sno} clusters with the {lcao}.
By using generated a seed image, this upgrade process allows you to skip multiple minor releases and it is faster than the standard upgrade method.

You generate the seed image from a cluster that is upgraded and configured with the desired {product-title} and applications by using the standard upgrade method. This is the seed cluster.

[IMPORTANT]
====
The image-based upgrade uses custom images that are specific to the hardware platform the clusters are running on.
Each different hardware platform requires a separate custom image.
====

> With this seed image, you can upgrade the platform version on any {sno} that matches the image's cluster profile and then restore applications.

Author: That is not clear, you're right. Anyway, how about:
> You can upgrade the platform version on any {sno} that matches the cluster profile of the seed image.

> //How do we specify the hardware exactly? Is it baked into the seed image based on the seed cluster we're using?

Reviewer 9: The server model & hardware must be the same for the seed cluster as for the clusters we want to upgrade using the seed image created. Note that it is not mandatory to be the same server, but it must be the same hardware architecture, cpus, network interfaces... This is usually very clear with telco customers where they use exactly the same servers for edge deployments.

The {lcao} uses two custom resources (CRs):

> * The `SeedGenerator` CR specifies the platform version of the seed cluster that you want to use to upgrade the target clusters.

Reviewer 9: well, it specifies the name of the container image that is going to be created based on the seed cluster configuration. This container image will eventually be pushed to the registry once the image is completed.

example:

~~~yaml
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage
  namespace: openshift-lifecycle-agent
spec:
  seedImage: quay.io/alosadag/ibu-seed-sno0:v4.14.1
~~~

It will be pushed to my quay.io account. However, the naming depends on the user. I recommend <seed-cluster-name>:<ocp_version>.

* The `imageBasedUpgrade` CR specifies the generated seed image's location and you use it to step through the upgrade stages.

After generating the seed image, the upgrade process has four stages:

* `Idle`
* `Prep`
* `Upgrade`
* `Rollback` (Optional)

> 

Reviewer 9: The IBU CR is created automatically once the LCA operator is installed. By default, the IBU spec only contains stage: Idle IIRC.  So, that's the first stage of the process.

Reviewer: Idle is the default stage

[discrete]
== Idle stage

> There is no ongoing upgrade and the cluster is ready to move to the `Prep` stage.

Reviewer 9: Default stage when the LCA operator is installed.

[discrete]
== Prep stage

> You can complete this stage in advance.

Reviewer 9: I do not understand that sentence. Do you mean this stage has some prerequisites?

Author: No, I meant to say that you can complete this stage outside of and regardless of the maintenance window ( if I understand it correctly).
I saw it phrased similarly as I did in one of the docs, but I'll clarify this.

Reviewer: yes

The {lcao} ensures that the node is ready to proceed to the `Upgrade` stage. This includes the following steps:

> * Installing and configuring the OADP Operator

Reviewer 9: As far as I know, the OADP operator must be installed by the user previously cc @donpenney @browsell . It can be installed in many ways depending on the cluster, but in our case where we foresee a fleet of clusters being managed by RHACM, the ZTP/TALM creating the proper policies makes more sense to me.

Reviewer: it is a prerequisite that the oadp operator is installed on the target cluster prior to starting this workflow

> * Setting up the storage bucket for backup

Reviewer 9: same as before, the way it is written looks like the Prep stage sets the storage solution for you.  Notice that an Object storage S3 solution is required by OADP.

> * Precaching list of images from the seed image

Reviewer 9: The precaching list of images to be pulled from the target cluster is included as a list in the seed image itself.  With that list, the target cluster runs a precache Job to pull all of them. If the precache fails the whole process is stopped/fails. 

Yes, one of the steps of the Prep stage is to pull in advance the container images that were stored in the seed cluster (except the catalogsources images or catalog index images).

> The {lcao} uses the `ostree` versioning system and the OADP Operator to create a backup, which allow complete cluster reconfiguration after upgrade and rollback.

Reviewer 9: well, the ostree at 10000 feet is used to create a copy of the seed cluster in a container image. The OADP operator as you stated is used by LCA to backup and restore what is needed to run your applications once the cluster is upgraded.

At this stage, the {lcao} pulls the seed image to the target cluster and two stateroots are created by `ostree`:

* stateroot of the current version of the target cluster
* stateroot of the newer version that is based on the seed image

> 

Reviewer 9: Example. See that the default stateroot is called rhcos while the one created by LCA is currently named as rhcos_$ocp_version

~~~sh
[root@snonode ~]# rpm-ostree status -v

State: idle
AutomaticUpdates: disabled
Deployments:  ostree-unverified-registry:quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:4c35b047a35eee940f136034f31c72d08c171da6e4d1ac73dcfee322d9659777 (index: 0)
                   Digest: sha256:4c35b047a35eee940f136034f31c72d08c171da6e4d1ac73dcfee322d9659777
                  Version: 414.92.202312011602-0 (2024-01-15T14:19:36Z)
                   Commit: 7e2137d179196775acf74ffe95150c1fa984b620b77da5c2c211bdcab94b8ab6
                   Staged: no
                StateRoot: rhcos_4.14.6
● ostree-unverified-registry:quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f4ad0c56ede0a79651cf618601ab0bb93d64a8a15d3d6f82ccb944d6aded7735 (index: 1)
                   Digest: sha256:f4ad0c56ede0a79651cf618601ab0bb93d64a8a15d3d6f82ccb944d6aded7735
                  Version: 414.92.202310270216-0 (2024-01-11T16:12:12Z)
                   Commit: da48dfaea37c2f3380a5d1c1e1b673c68184a9c34581d5fada4833e495dfe04a
                StateRoot: rhcos
~~~

In a future stage (Upgrade IIRC) the default stateroot when booting will be the rhcos_$ocp_version.

Reviewer: Too much internals for an end user doc. Admin will initiate the Prep stage by patching the CR and should monitor the status of the CR

You can stop the upgrade process at this stage. In that case, the Operator performs cleanup operations and sets the upgrade stage back to `Idle`.

> 

Reviewer 9: Moving to Idle will do the cleanup. Not sure if it is important to notice that the OADP operator will be also removed, so the user will need to install it again. A little bit annoying personally.

Reviewer: Please check the lastest code base. It is not removed

Reviewer 9: Still, it is not clear to me that sentence. It is the action to set IBU spec.stage to Idle that it will stop the upgrade process. Maybe it can be rephrased as:

```
You can stop the upgrade process by setting the stage back to `Idle` If you stop, the Operator performs cleanup operations.
```

[discrete]
== Upgrade stage

> Just before the {lcao} starts the upgrade process, the OADP Operator creates a backup of the artifacts on a compatible storage solution, such as an Amazon S3 bucket. These OADP `Backup` and `Restore` CRs are also copied to the new stateroot to be applied after the upgrade

Reviewer 9: I would encourage to add as an example Ceph Object Gateway which is compatible with Amazon S3. In my case I've been using MinIO which is an OCP certified operator IIRC.

Reviewer: We should not be dictating the backend

Once the upgrade is successful, the Operator sets the default stateroot to the new one and applies the cluster and application configurations.

> The Operator also regenerates the cluster's cryptographic certifications and keys.

Reviewer 9: cluster's cryptographic certificates?

> This ensures that each {sno} upgraded with the same seed image has unique and valid cryptographic object.

Reviewer 9: I think it should be objects since many certificates need to be generated to target cluster's IP, hostname, etc...

[discrete]
== Rollback stage (Optional)

> You can use the backup from the `Upgrade` stage to start a rollback if the upgrade fails

Reviewer 9: fails. (dot)

If you roll back the upgrade, the {lcao} sets the default stateroot to the previous version and the cluster boots up with the previous release of {product-title} and applications configurations.

> After a successful upgrade or a rollback, you commit to the change by moving the stage back to `Idle`.

Reviewer 9: should be warned that this is a manual step or a step to be executed by the user. I mean, LCA won't do that after a successful update. Keep honest @donpenney

Author: You're right, it's a manual step, I'll rephrase to clarify that.
